p2p/dandelion/src/router.rs

```diff
@@ -0,0 +1,384 @@
+    pub(crate) outbound_peer_discover: Pin<Box<P>>,
+    /// A [`Service`] which handle broadcasting (diffusing) transactions.
+    pub(crate) broadcast_svc: B,
+
+    /// The current state.
+    pub(crate) current_state: State,
+    /// The time at which this epoch started.
+    pub(crate) epoch_start: Instant,
+
+    /// The stem our local transactions will be sent to.
+    pub(crate) local_route: Option<ID>,
+    /// A [`HashMap`] linking peer's IDs to IDs in `stem_peers`.
+    pub(crate) stem_routes: HashMap<ID, ID>,
+    /// Peers we are using for stemming.
+    ///
+    /// This will contain peers, even in [`State::Fluff`] to allow us to stem [`TxState::Local`]
+    /// transactions.
+    pub(crate) stem_peers: HashMap<ID, S>,
+
+    /// The distribution to sample to get the [`State`], true is [`State::Fluff`].
+    pub(crate) state_dist: Bernoulli,
+
+    /// The config.
+    config: DandelionConfig,
+
+    /// The routers tracing span.
+    span: tracing::Span,
+
+    _tx: PhantomData<Tx>,
+}
+
+impl<Tx, ID, P, B, S> DandelionRouter<P, B, ID, S, Tx>
+where
+    ID: Hash + Eq + Clone,
+    P: Discover<Key = ID, Service = S, Error = tower::BoxError>,
+    B: Service<DiffuseRequest<Tx>, Error = tower::BoxError>,
+    S: Service<StemRequest<Tx>, Error = tower::BoxError>,
+{
+    /// This function gets the number of outbound peers from the [`Discover`] required for the selected [`Graph`](crate::Graph).
+    fn poll_prepare_graph(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), DandelionRouterError>> {
+        let peers_needed = match self.current_state {
+            State::Stem => self.config.number_of_stems(),
+            // When in the fluff state we only need one peer, the one for our txs.
+            State::Fluff => 1,
+        };
+
+        while self.stem_peers.len() < peers_needed {
+            match ready!(self
+                .outbound_peer_discover
+                .as_mut()
+                .poll_discover(cx)
+                .map_err(DandelionRouterError::OutboundPeerDiscoverError))
+            .ok_or(DandelionRouterError::OutboundPeerDiscoverExited)??
+            {
+                Change::Insert(key, svc) => {
+                    self.stem_peers.insert(key, svc);
+                }
+                Change::Remove(key) => {
+                    self.stem_peers.remove(&key);
+                }
+            }
+        }
+
+        Poll::Ready(Ok(()))
+    }
+
+    fn fluff_tx(&mut self, tx: Tx) -> B::Future {
+        self.broadcast_svc.call(DiffuseRequest(tx))
+    }
+
+    fn stem_tx(&mut self, tx: Tx, from: ID) -> S::Future {
+        loop {
+            let stem_route = self.stem_routes.entry(from.clone()).or_insert_with(|| {
+                self.stem_peers
+                    .iter()
+                    .choose(&mut thread_rng())
+                    .expect("No peers in `stem_peers` was poll_ready called?")
+                    .0
+                    .clone()
+            });
```

Comment:
Maybe, I know this is very inefficient but the maximum amount of peers is 2 and this will only be ran once every 10 mins per inbound connection. So if you know a way I can improve it I will, but this shouldn't be an issue.